src/signatures.rs

```diff
@@ -0,0 +1,1206 @@
+            let map = match &on_chain_cred.value {
+                AccountCredentialWithoutProofs::Initial { icdv } => &icdv.cred_account.keys,
+                AccountCredentialWithoutProofs::Normal { cdv, .. } => &cdv.cred_key_info.keys,
+            };
+
+            if !map.contains_key(&KeyIndex(*inner_key)) {
+                return Err(SignatureError::MissingIndicesOnChain {
+                    credential_index: *outer_key,
+                    key_index:        *inner_key,
+                });
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Verify a given message was signed by the given account.
+/// Concretely this means:
+/// - enough credential holders signed the message.
+/// - each of the credential signatures has the required number of signatures.
+/// - all of the signatures are valid, that is, it is not sufficient that a
+///   threshold number are valid, and some extra signatures included are
+///   invalid.
+pub async fn verify_account_signature(
+    mut client: v2::Client,
+    signer: AccountAddress,
+    signatures: &AccountSignatures,
+    message: &Message<'_>,
+    bi: BlockIdentifier,
+) -> Result<bool, SignatureError> {
+    let message_hash = calculate_message_hash(message, signer);
+
+    let signer_account_info = client
+        .get_account_info(&AccountIdentifier::Address(signer), bi)
+        .await?;
+
+    let signer_account_credentials = signer_account_info.response.account_credentials;
+    let credential_signatures_threshold = signer_account_info.response.account_threshold;
+
+    // Ensure that all key indices in the signatures map exist on chain but not vice
+    // versa. This allows us to iterate through the `signer_account_credentials` map
+    // to get the public keys and thresholds from the on-chain information
+    // further down while still knowing that each signature in the signatures map
+    // has a corresponding entry on chain.
+    assert_signature_map_key_indices_on_chain(signatures, &signer_account_credentials)?;
+
+    let mut valid_credential_signatures_count = 0u8;
+    for (credential_index, credential) in signer_account_credentials {
+        // Retrieve the public key and threshold from the on-chain information.
+        let (keys, signatures_threshold) = match credential.value {
+            AccountCredentialWithoutProofs::Initial { icdv } => {
+                (icdv.cred_account.keys, icdv.cred_account.threshold)
+            }
+            AccountCredentialWithoutProofs::Normal { cdv, .. } => {
+                (cdv.cred_key_info.keys, cdv.cred_key_info.threshold)
+            }
+        };
+
+        let mut valid_signatures_count = 0u8;
+
+        for (key_index, public_key) in keys {
+            // If a signature exists for the given credential and key index, verify it and
+            // increase the `valid_signatures_count`.
+            if let Some(cred_sigs) = signatures.sigs.get(&credential_index.index) {
```

Comment:
Consider `let Some(cred_sigs) = signatures.sigs.get(&credential_index.index) else { continue; }` to avoid the indentation.